dpe/src/x509.rs

```diff
@@ -1088,6 +1299,251 @@ impl X509CertWriter<'_> {
+        bytes_written += self.encode_ueid(measurements)?;
+        bytes_written += self.encode_basic_constraints(measurements)?;
+        bytes_written += self.encode_key_usage(measurements.is_ca)?;
+        bytes_written += self.encode_extended_key_usage(measurements)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode a SignedData
+    ///
+    /// This function does not populate the certificates or crls fields.
+    ///
+    /// SignedData  ::=  SEQUENCE  {
+    ///    version CMSVersion,
+    ///    digestAlgorithms DigestAlgorithmIdentifiers,
+    ///    encapContentInfo EncapsulatedContentInfo,
+    ///    certificates [0] IMPLICIT CertificateSet OPTIONAL,
+    ///    crls [1] IMPLICIT RevocationInfoChoices OPTIONAL,
+    ///    signerInfos SignerInfos
+    /// }
+    #[allow(clippy::identity_op)]
+    fn encode_signed_data(
+        &mut self,
+        serial_number: &[u8],
+        issuer_name: &[u8],
+        csr: &[u8],
+        sig: &EcdsaSig,
+    ) -> Result<usize, DpeErrorCode> {
+        // SignedData is EXPLICIT field number 0
+        let mut bytes_written =
+            self.encode_byte(Self::CONTEXT_SPECIFIC | Self::CONSTRUCTED | 0x0)?;
+        bytes_written += self.encode_size_field(Self::get_signed_data_size(
+            csr,
+            serial_number,
+            issuer_name,
+            sig,
+            /*tagged=*/ true,
+            /*explicit=*/ false,
+        )?)?;
+
+        // SignedData sequence
+        bytes_written += self.encode_tag_field(Self::SEQUENCE_TAG)?;
+        bytes_written += self.encode_size_field(Self::get_signed_data_size(
+            csr,
+            serial_number,
+            issuer_name,
+            sig,
+            /*tagged=*/ false,
+            /*explicit=*/ false,
+        )?)?;
+
+        // CMS version
+        bytes_written += self.encode_integer(Self::CMS_V1)?;
+
+        // digestAlgorithms
+        bytes_written += self.encode_tag_field(Self::SET_OF_TAG)?;
+        bytes_written +=
+            self.encode_size_field(Self::get_hash_alg_id_size(/*tagged=*/ true)?)?;
+        bytes_written += self.encode_hash_alg_id()?;
+
+        // encapContentInfo
+        bytes_written += self.encode_encapsulated_content_info(csr)?;
+
+        // signerInfos
+        bytes_written += self.encode_tag_field(Self::SET_OF_TAG)?;
+        bytes_written += self.encode_size_field(Self::get_signer_info_size(
+            serial_number,
+            issuer_name,
+            sig,
+            /*tagged=*/ true,
+        )?)?;
+        bytes_written += self.encode_signer_info(serial_number, issuer_name, sig)?;
+
+        Ok(bytes_written)
+    }
+
+    /// Encode an attributes structure
+    ///
+    /// Attributes ::= SET OF Attribute
+    ///
+    /// Attribute ::= SEQUENCE {
+    ///    attrType OBJECT IDENTIFIER,
+    ///    attrValues SET OF AttributeValue
+    /// }
+    ///
+    /// AttributeValue ::= ANY -- Defined by attribute type
+    #[allow(clippy::identity_op)]
```

Comment:
Why is this needed? Because of the `| 0x0`?

Reply:
Yes